packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,224 @@
-): JSONArray => {
-  const argTypes = match(arrow.domain)
-    .with({ tag: "labeledProduct" }, (x) => {
-      return Object.values(x.fields);
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      return x.items;
-    })
-    .with({ tag: "nil" }, (x) => {
-      return [];
-    })
-    // uncomment to check that every pattern in matched
-    // .exhaustive()
-    .otherwise(() => {
-      throw new Error("Unexpected tag: " + jsonify(arrow.domain));
+    return value.map((y, i) => {
+      return aqua2ts(y, schema.items[i]);
     });
-
-  if (req.args.length !== argTypes.length) {
-    throw new Error(
-      `incorrect number of arguments, expected: ${argTypes.length}, got: ${req.args.length}`,
-    );
+  } else if (["labeledProduct", "struct"].includes(schema.tag)) {
+    if (typeof value !== "object" || value == null || Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return Object.entries(schema.fields).reduce((agg, [key, type]) => {
+      const val = aqua2ts(value[key], type);
+      return { ...agg, [key]: val };
+    }, {});
+  } else {
+    throw new Error("Unexpected tag: " + JSON.stringify(schema));
   }
-
-  return req.args.map((arg, index) => {
-    return aqua2ts(arg, argTypes[index]);
-  });
-};
-
-/**
- * Convert value from its typescript representation to representation in aqua
- * @param value - the value as represented in typescript
- * @param type - definition of the aqua type
- * @returns value represented in aqua
- */
-export const ts2aqua = (value: JSONValue, type: NonArrowType): JSONValue => {
-  const res = match(type)
-    .with({ tag: "nil" }, () => {
-      return null;
-    })
-    .with({ tag: "option" }, (opt) => {
-      if (value === null || value === undefined) {
-        return [];
-      } else {
-        return [ts2aqua(value, opt.type)];
-      }
-    })
-    .with({ tag: "scalar" }, { tag: "bottomType" }, { tag: "topType" }, () => {
-      return value;
-    })
-    .with({ tag: "array" }, (arr) => {
-      assert(Array.isArray(value), "Should not be possible, bad types");
-      return value.map((y) => {
-        return ts2aqua(y, arr.type);
-      });
-    })
-    .with({ tag: "struct" }, (x) => {
-      return Object.entries(x.fields).reduce((agg, [key, type]) => {
-        const val = ts2aqua(value[key], type);
-        return { ...agg, [key]: val };
-      }, {});
-    })
-    .with({ tag: "labeledProduct" }, (x) => {
-      return Object.entries(x.fields).reduce((agg, [key, type]) => {
-        const val = ts2aqua(value[key], type);
-        return { ...agg, [key]: val };
-      }, {});
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      return x.items.map((type, index) => {
-        return ts2aqua(value[index], type);
-      });
-    })
-    // uncomment to check that every pattern in matched
-    // .exhaustive()
-    .otherwise(() => {
-      throw new Error("Unexpected tag: " + jsonify(type));
+}
+
+export function ts2aqua(
```

Comment:
I would also maybe name this `jsToAqua` cause it really doesn't do anything related specifically to ts